src/GafferUI/GraphGadget.cpp

```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
 	return nullptr;
 }
 
+void GraphGadget::dirtyActive()
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+	m_activeStateDirty = true;
+	dirty( DirtyType::Render );
+}
+
+void GraphGadget::updateActive() const
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+
+	tbb::spin_mutex::scoped_lock scopedLockA( m_pendingActiveStateMutex );
+	const Gaffer::Node *focusNode = m_scriptNode->getFocus();
+	std::vector<Gaffer::ConstPlugPtr> focusPlugs;
+	if( focusNode )
+	{
+		Gaffer::ConstPlugPtr hiddenFocusPlug = nullptr;
+		for( auto &plug : Gaffer::Plug::OutputRange( *focusNode ) )
+		{
+			const std::string &n = plug->getName();
+			if( n.size() >= 2 && n[0] == '_' && n[1] == '_' )
+			{
+				if( !hiddenFocusPlug )
+				{
+					hiddenFocusPlug = plug;
+				}
+				continue;
+			}
+
+			focusPlugs.push_back( plug );
+		}
+
+		if( !focusPlugs.size() && hiddenFocusPlug )
+		{
+			// If we couldn't find any visible output plug, take the hidden one
+			focusPlugs.push_back( hiddenFocusPlug );
+		}
+	}
+
+	if( !focusPlugs.size() )
+	{
+		m_pendingActiveState.reset();
+		scopedLockA.release();
+		Gaffer::ParallelAlgo::callOnUIThread(
+			[this] {
+				this->applyActive( false );
+			}
+		);
+		return;
+	}
+
+
+	tbb::spin_mutex &pendingActiveStateMutex = m_pendingActiveStateMutex;
+	std::unique_ptr<PendingActive> &pendingActiveState = m_pendingActiveState;
+
+	// Must hold a reference to stop us dying before our UI thread call is scheduled.
+	ConstGraphGadgetPtr thisRef = this;
+
+	m_activeStateTask = Gaffer::ParallelAlgo::callOnBackgroundThread(
+		// Subject:  TODO - does it matter that we're only including one plug as a subject?  We respond to any
+		// plugs being dirtied on the focus node with manual cancellation of this task, so hopefully it should
+		// be OK
+		focusPlugs[0].get(),
+		[focusPlugs, &pendingActiveStateMutex, &pendingActiveState, thisRef ] {
+			// We don't use the current context for our actual context - we use the script context
+			// and activePlugsAndNodes needs to continue even in cases where it can't figure out an accurate
+			// context andt just uses null ... but we do need to take the canceller from the context
+			// provided by callOnBackgroundThread
+			const IECore::Canceller *canceller = Gaffer::Context::current()->canceller();
```

Comment:
Where do we transfer the canceller into the contexts used to evaluate the graph? Assuming we're not doing it yet, perhaps the `EditableScope::setCanceller()` method I just added might be useful? (https://github.com/GafferHQ/gaffer/pull/4368/commits/5de5ccdfb2b77c453e0c981a8325862be5bc535e)

Reply:
Now that I've removed the canceller argument from the API for activePlugsAndNodes, and activeWalk is an anonymous internal function, there's no way to get here without the canceller already being in the context used to evaluate the graph.

While thinking about this further though, I realized that I haven't really confirmed that a useful Canceller is actually in the Context that updateActive gets called with.  Upon testing, I don't see any sign that cancellation is actually doing anything.  So that probably needs looking into, though I'm trying to ignore that for now and just get everything cleaned up.

Reply:
I did poke a bit further, and it's now working fine.  There were two issues:  I had accidentally called `new Context( context, otherContext->canceller() )`, not realizing that without dereferencing the Canceller pointer, it gets converted to bool or something.  The other thing I ran into was that my test graph used an OSL expression to connect things, and after fixing that, I got a crash, which is fixed by: 804729b

Reply:
Canceller passing is updated in a9b4961a2